### PR TITLE
[Merged by Bors] - feat: the upstreamDecl linter

### PR DESCRIPTION
**Pull request by adomani**
This linter is intended to work much like `minImports` in helping with splitting files. It places a warning on each set of declarations that could be moved to an earlier file. It takes into account dependencies on declarations in the current file, so that a whole block of declarations can be moved even if there are dependencies between the declarations in the block.

Since definitions are often deliberately placed in their own file, this linter does not place a warning on any declaration depending on a definition in the current file.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

**Comment by adomani** [comment]
maintainer delegate

**Comment by Vierkantor** [comment]
I have done some cleanup and I think this is good to go!

**Review by adomani — COMMENTED** [review]
Thank you very much for the improvements!

I only left a few small suggestions, but I am happy with this!

I am a bit hesitant to maintainer merge this, since I was very involved in the initial development, but I think that this is very useful!

**Comment by adomani** [comment]
One more comment: the linter is called `upstreamableDecl`, but the PR title uses `upstreamDecl`.  I may actually have a slight preference for `upstreamDecl` as the name of the linter as well.

**Comment by kim-em** [comment]
@Vierkantor, could you address @adomani's remaining suggestions, and then merge?

**Comment by kim-em** [comment]
bors d+

**Comment by adomani** [comment]
I think that Kim meant

bors delegate=Vierkantor

😊

File: Mathlib/Tactic/Linter/UpstreamableDecl.lean [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+      return
+    if stx == (← `(command| set_option $(mkIdent `linter.upstreamableDecl) true)) then return
+    let env ← getEnv
+    let id ← getId stx
+    if id != .missing then
+      let minImports := getIrredundantImports env (← getAllImports stx id)
+      match minImports with
+      | ⟨(RBNode.node _ .leaf upstream _ .leaf), _⟩ => do
+        if !(← env.localDefDependencies stx id) then
+          logWarning f!"Consider moving this declaration to the module {upstream}."
```
[the remark this answers is not in the document]
Comment (Vierkantor) [reply Mathlib/Tactic/Linter/UpstreamableDecl.lean]: I used the `logLint` and the widget, thanks for teaching me that! One thing though, looks like neovim doesn't render this widget specially (I think they need to manually implement support for each widget). Could you check that it works okay in VS Code?

[the remark this answers is not in the document]
Comment (adomani) [reply Mathlib/Tactic/Linter/UpstreamableDecl.lean]: I just checked in VSCode and it works great: the module name is clickable (and clearly identifiable as a link) and clicking on it opens the corresponding module.

Thanks!

**Comment by kim-em** [comment]
bors merge

**Pull request closed without merging**